### PR TITLE
Exposed enrollment date fields on course run form

### PR DESCRIPTION
**Pull request by clintonb**
LEARNER-2621

### CODE REVIEW

**Comment by clintonb** [comment]
screen shot 2017-09-25 at 4 15 16 pm

FYI @srpearce I have added enrollment dates to the course run edit page.

**Comment by srpearce** [comment]
@clintonb Other dates on the Course Run page aren't grouped this way (see screen shot below). I personally prefer it, and think the course start and end dates should be grouped as well, but I don't know if there was a reason to separate them or if there are guidelines to follow around UI consistency. @roderickmorales, can you weigh in?

screenshot 5

**Comment by roderickmorales** [comment]
@clintonb we should meet and let me know what your plan is for changes. I'm having a hard time understanding why adding these dates has more priority than other things that need fixing in Publisher.

Also, as this seems like an opportunity to address some of the things that Sylvia mentioned above

**Comment by clintonb** [comment]
Publisher is _not_ my priority. We are only touching Publisher in order to have a single editor for course details. Our goal is to power the course detail pages on the marketing site with the Discovery Service (rather than Drupal). Publisher is the best way to get data into the Discovery Service (since both share a database).

If there are no glaring issues, I would like to get this merged soon so that we can move forward with our Drupal modifications.

File: course_discovery/apps/publisher/forms.py [hunk 1]
```diff
@@ -323,30 +324,48 @@ def clean_lms_course_id(self):
...
+
+        if start and end and start >= end:
+            self.add_error('start', _('The start date must occur before the end date.'))
+
+        if min_effort and max_effort and min_effort >= max_effort:
+            self.add_error('min_effort', _('Minimum effort must be less than maximum effort.'))
+
         if is_xseries and not xseries_name:
-            raise ValidationError({'xseries_name': _('Enter XSeries program name')})
+            self.add_error('xseries_name', _('Please provide the name of the associated XSeries.'))
```
Comment (srpearce) [inline course_discovery/apps/publisher/forms.py:345]: Enter the name of the XSeries program.

File: course_discovery/apps/publisher/forms.py [hunk 2]
```diff
@@ -323,30 +324,48 @@ def clean_lms_course_id(self):
...
+
         if is_micromasters and not micromasters_name:
-            raise ValidationError({'micromasters_name': _('Enter Micromasters program name')})
+            self.add_error('micromasters_name', _('Please provide the name of the associated MicroMasters.'))
```
Comment (srpearce) [inline course_discovery/apps/publisher/forms.py:348]: Enter the name of the MicroMasters program.

File: course_discovery/apps/publisher/forms.py [hunk 3]
```diff
@@ -323,30 +324,48 @@ def clean_lms_course_id(self):
...
+
         if is_professional_certificate and not professional_certificate_name:
-            raise ValidationError({'professional_certificate_name': _('Enter Professional Certificate program name')})
+            self.add_error('professional_certificate_name',
+                           _('Please provide the name of the associated Professional Certificate program.'))
```
Comment (srpearce) [inline course_discovery/apps/publisher/forms.py:352]: Enter the name of the Professional Certificate program.

File: course_discovery/apps/publisher/forms.py [hunk 4]
```diff
@@ -413,18 +432,18 @@ def clean(self):
         credit_price = self.cleaned_data.get('credit_price')
         seat_type = self.cleaned_data.get('type')
 
-        if seat_type in [Seat.PROFESSIONAL, Seat.VERIFIED, Seat.CREDIT] and not price:
-            self.add_error('price', _('Only audit seat can be without price.'))
+        if seat_type and seat_type != Seat.AUDIT and not price:
+            self.add_error('price', _('Please specify a price.'))
```
Comment (srpearce) [inline course_discovery/apps/publisher/forms.py:436]: Specify a price.

File: course_discovery/apps/publisher/forms.py [hunk 5]
```diff
@@ -413,18 +432,18 @@ def clean(self):
...
 
         if seat_type == Seat.CREDIT and not credit_price:
-            self.add_error('credit_price', _('Only audit seat can be without price.'))
+            self.add_error('credit_price', _('Please specify a price for credit.'))
```
Comment (srpearce) [inline course_discovery/apps/publisher/forms.py:439]: Specify a price for course credit.

File: course_discovery/apps/publisher/templates/publisher/course_run/edit_run_form.html [hunk 6]
```diff
@@ -269,6 +297,53 @@ <h1 class="hd-1 emphasized">
...
                 </div>
 
+            {% switch "display_enrollment_dates_on_publisher_form" %}
+                <div class="field-title">{% trans "Enrollment Dates" %}</div>
+                <div class="row">
+                    <div class="col col-6 help-text">
+                        {% blocktrans trimmed %}
+                            The date on which the course opens for enrollment. This date must occur on or before
+                            the course start date. Leaving this field empty will result in the default value,
+                            the course start date, being used.
```
Comment (srpearce) [inline course_discovery/apps/publisher/templates/publisher/course_run/edit_run_form.html:307]: The date enrollment starts for the course. This date must be on or before the course start date. If you leave this field blank, enrollment starts on the course start date.

File: course_discovery/apps/publisher/templates/publisher/course_run/edit_run_form.html [hunk 7]
```diff
@@ -269,6 +297,53 @@ <h1 class="hd-1 emphasized">
...
+                            </div>
+                        {% endif %}
+                    </div>
+                </div>
+                <div class="row">
+                    <div class="col col-6 help-text">
+                        {% blocktrans trimmed %}
+                            The date on which course enrollment is closed. This date must occur after the course start
+                            date, but before the course end date. Leaving this field empty will result in the default value,
+                            the course end date, being used.
```
Comment (srpearce) [inline course_discovery/apps/publisher/templates/publisher/course_run/edit_run_form.html:329]: The date enrollment ends for the course. This date must be after the course start date, but before the course end date. If you leave this field blank, enrollment ends on the course end date.

File: course_discovery/apps/publisher/forms.py [hunk 8]
```diff
@@ -323,30 +324,48 @@ def clean_lms_course_id(self):
...
+                self.add_error('enrollment_start',
+                               _('The enrollment start date must occur on or before the course start date.'))
+        else:
+            cleaned_data['enrollment_start'] = start
+
+        enrollment_end = cleaned_data.get('enrollment_end')
+        if enrollment_end:
+            if end and enrollment_end > end:
+                self.add_error('enrollment_end',
+                               _('The enrollment end date must occur on or after the course end date.'))
```
Comment (srpearce) [inline course_discovery/apps/publisher/forms.py:366]: The enrollment end date must be after the enrollment start date, but before the course end date.

File: course_discovery/conf/locale/en/LC_MESSAGES/django.po [hunk 9]
```diff
@@ -737,31 +737,33 @@ msgid "Video Language"
...
 msgstr ""
 
 #: apps/publisher/forms.py
-msgid "Minimum effort cannot be greater than Maximum effort"
+msgid "Minimum effort must be less than maximum effort."
 msgstr ""
 
 #: apps/publisher/forms.py
-msgid "Minimum effort and Maximum effort can not be same"
+msgid "Please provide the name of the associated XSeries."
```
Comment (srpearce) [inline course_discovery/conf/locale/en/LC_MESSAGES/django.po:748]: See the changes for this and other fields earlier in this PR.

**Comment by clintonb** [comment]
We have opted *not* to expose these fields to the end user. If they need to be modified, PCs will do so via Django admin.

**Pull request closed without merging**